app/services/brapi/v1/germplasm_queries.rb

```diff
@@ -122,17 +139,23 @@ def execute_statement(sql, atts)
     # at least until the previous one has been deallocated
     Thread.exclusive do
       @connection.prepare('brapi_statement', sql)
-      
-      if( (atts != nil) && !(atts.empty?) )
-        results = @connection.exec_prepared("brapi_statement", atts)
-      else
-        results = @connection.exec_prepared("brapi_statement")
+      begin
+        if( (atts != nil) && !(atts.empty?) )
+          results = @connection.exec_prepared("brapi_statement", atts)
+        else
+          results = @connection.exec_prepared("brapi_statement")
+        end
+      rescue PG::Error => e
+        @connection.exec("ROLLBACK") 
+        results = nil
+        Rails.logger.warn { "Encountered an error executing a BrAPI germplasm-related query: #{e.message} #{e.backtrace.join("\n")}" }
```

Comment:
There are several issues with this approach. Firstly, the query failing in such way is not something we can ignore, this is clearly error in application logic and we want to know about it (well, it could be a timeout but we still want to know about it). We use mail notifier which sends information about exceptions that are not handled at the controller level ([smartinez/exception_notification](https://github.com/smartinez87/exception_notification) gem). However it will not run if the exception is handled by application code. In some such cases it makes sense to call it directly:
```
ExceptionNotifier.notify_exception(e)
```

Secondly, returning 404 response in case of application error is misleading. 500 should be returned so that the user knows this is in fact an error and not that we have no information he/she requested. 

Lastly, returning `nil` from method to indicate error may not be terribly wrong but it could be done in more explicit way, e.g.: do cleanup, log debugging info, send error notification and return `:error` instead of `nil` or alternatively do cleanup and re-raise the exception and use error handler at the controller level (`rescue_from PG::Error`). Optionally, wrap `PG::Error` with custom `QueryError` to avoid leaky abstraction. Where possible error handling should be separated from other logic as it tends to make code less readable.

```ruby
class Brapi::V1::GermplasmQueries
# ...
  rescue PG::Error => ex
    @connection.exec("ROLLBACK") 
    raise QueryError, sql # or whatever useful information (note that previous exception is automatically attached as the cause of new exception)
  ensure
    @connection.exec("DEALLOCATE brapi_statement") 
  end
# ...
```

```ruby
class Brapi::V1::ApidocsController < ApplicationController
# ...
  rescue_from Brapi::QueryError do |ex|
    logger.warn { ... }
    ExceptionNotifier.notify_exception(ex, env: request.env, data: { sql: ex.sql, cause: ex.cause.message })
    return json: { reason: 'Internal error', message: 'There was some error managing studies/search query' }, status: :internal_server_error
  end
# ...
```